include/stdx/latched.hpp

```diff
@@ -0,0 +1,61 @@
+#pragma once
+
+#include <stdx/compiler.hpp>
+#include <stdx/functional.hpp>
+#include <stdx/type_traits.hpp>
+
+#include <optional>
+#include <type_traits>
+#include <utility>
+
+namespace stdx {
+inline namespace v1 {
+template <typename F> struct latched {
+    using value_type = stdx::remove_cvref_t<std::invoke_result_t<F>>;
+
+    constexpr explicit latched(F const &f) : lazy{f} {}
+    constexpr explicit latched(F &&f) : lazy{std::move(f)} {}
+
+    constexpr auto has_value() const noexcept -> bool {
+        return opt.has_value();
+    }
+    constexpr explicit operator bool() const noexcept {
+        return opt.has_value();
+    }
+
+    constexpr auto value() const & LIFETIMEBOUND -> value_type const & {
+        populate();
+        return *opt;
+    }
+    constexpr auto value() const && LIFETIMEBOUND -> value_type const && {
+        populate();
+        return *std::move(opt);
+    }
+
+    constexpr auto operator->() const LIFETIMEBOUND->value_type const * {
+        populate();
+        return opt.operator->();
+    }
+
+    constexpr auto operator*() const & LIFETIMEBOUND->decltype(auto) {
+        return value();
+    }
+    constexpr auto operator*() const && LIFETIMEBOUND->decltype(auto) {
+        return std::move(*this).value();
+    }
+
+  protected:
+    constexpr auto populate() const {
+        if (not opt.has_value()) {
+            opt.emplace(lazy);
```

Comment:
question:
Should this be 
```c++
opt.emplace(lazy());`
```

Reply:
This code is unchanged, just copied. And no -- calling the function explicitly here would incur a move operation. The point of lazy is to be able to emplace non-movable things, which it does by implicit conversion to `T`: the function is called at the time of conversion.

Reply:
I didn't look into `with_result_of<F>`. Very nice!